### PR TITLE
Implement `CustomDiffusionAttnProcessor2_0`.

### LINKED ISSUE #4588: Implement `CustomDiffusionAttnProcessor2_0`
Default AttnProcessor and LoRA can use torch 2.0 SDP operators, but the custom diffusion processors can not.
I can start a PR if you think it is good.

### PR DESCRIPTION
**Pull request by eliphatfs**
# What does this PR do?

Fixes #4588

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@sayakpaul @nupurkmr9

### CODE REVIEW

**Comment by sayakpaul** [comment]
@eliphatfs let me know if this PR is ready for review.

**Comment by eliphatfs** [comment]
Hi, how do I do the `ko`-documentation? I don't know any Korean.

**Comment by eliphatfs** [comment]
I don't find existing tests for CustomDiffusionAttnProcessor, so I will skip adding that for `2_0`.

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -1466,6 +1469,118 @@ def __call__(self, attn: Attention, hidden_states, encoder_hidden_states=None, a
...
+        hidden_states = F.scaled_dot_product_attention(
+            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
+        )
+
+        hidden_states = hidden_states.transpose(1, 2).reshape(batch_size, -1, attn.heads * head_dim)
+        hidden_states = hidden_states.to(query.dtype)
+
+        # attention_probs = attn.get_attention_scores(query, key, attention_mask)
+        # hidden_states = torch.bmm(attention_probs, value)
+        # hidden_states = attn.batch_to_head_dim(hidden_states)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:1568]: Can go away.

**Review by sayakpaul — COMMENTED** [review]
Looking good, thank you!

* Could we add this conditional initialization of this class in the training script? Here's an example: https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py#L876C1-L880C1. 


@yiyixuxu could you also give this a look?

**Comment by eliphatfs** [comment]
The CI complains about my code style. How do I spot the problem?

**Comment by sayakpaul** [comment]
You can run `make style && make quality` from you local fork of `diffusers`.

**Comment by eliphatfs** [comment]
Sorry I didn't read the guide carefully; I have updated the code formatting, but there are extra errors when I wrong `make quality` that seems nothing to do with me:
```
src/diffusers/pipelines/audio_diffusion/pipeline_audio_diffusion.py:181:12: E721 Do not compare types, use `isinstance()`
src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:793:42: E721 Do not compare types, use `isinstance()`
src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:878:20: E721 Do not compare types, use `isinstance()`
src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py:1083:20: E721 Do not compare types, use `isinstance()`
tests/pipelines/consistency_models/test_consistency_models.py:190:12: E721 Do not compare types, use `isinstance()`
tests/pipelines/unidiffuser/test_unidiffuser.py:112:12: E721 Do not compare types, use `isinstance()`
tests/pipelines/unidiffuser/test_unidiffuser.py:548:12: E721 Do not compare types, use `isinstance()`
tests/pipelines/unidiffuser/test_unidiffuser.py:651:12: E721 Do not compare types, use `isinstance()`
```

**Comment by sayakpaul** [comment]
Cc @DN6 could you help here?

**Comment by DN6** [comment]
@eliphatfs What version of `ruff` do you have installed?  You can check by running `ruff -V`. 
Could you share the output of that command please?

**Comment by eliphatfs** [comment]
Oh it was `0.0.284`; after installing `0.0.280` as specified the errors vanished. `0.0.280` was not installed on my machine due to some version clash in `flax` which refers to `orbax` package that is no longer installable: https://github.com/google/orbax/issues/436
I have fixed all style problems. Thanks for your help!!

**Comment by eliphatfs** [comment]
> Looking good, thank you!
> 
> * Could we add this conditional initialization of this class in the training script? Here's an example: https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py#L876C1-L880C1.
> 
> @yiyixuxu could you also give this a look?

Done!

File: src/diffusers/models/attention_processor.py (hunk 1 above)
Comment (eliphatfs) [reply src/diffusers/models/attention_processor.py]: Thanks and it goes away.

**Comment by eliphatfs** [comment]
Forgot to update loaders. Now `test_examples` pass on my local clone.

**Comment by sayakpaul** [comment]
@DN6 could you give this a look?

**Comment by eliphatfs** [comment]
Bump?

**Comment by sayakpaul** [comment]
We are doing two major refactors just as an FYI which might influence this PR:

https://github.com/huggingface/diffusers/pull/4473
https://github.com/huggingface/diffusers/pull/4765

Let's maybe revisit this after that? WDYT?

**Comment by eliphatfs** [comment]
I see the PRs you mentioned have been merged. Shall we continue working on this thing?

**Review by sayakpaul — APPROVED** [review]
Nice this works for me. Thank you so much for implementing this!

**Comment by sayakpaul** [comment]
Ah, we have some failures on the CI and merge conflicts. Let's make sure to resolve them.

**Comment by eliphatfs** [comment]
I have resolved the conflict; looking into the CI failures.

**Comment by eliphatfs** [comment]
Sorry but I request a re-run of the tests. It seemed to me that the errors are not relevant to this PR. CI is complaining about a field missing in a huggingface API result. I guess it may be due to API changes and since I have merged the changes in diffusers `main` branch maybe we can retry the tests.

**Review by patrickvonplaten — APPROVED** [review]
Ok for me. @sayakpaul @DN6 @yiyixuxu wdyt?

**Review by sayakpaul — APPROVED** [review]
Okay for me to merge once the CI is green (except the documentation workflow).

**Pull request merged**